genesyscloud/user_group_role_common.go

```diff
@@ -54,102 +54,142 @@ func getAssignedGrants(subjectID string, authAPI *platformclientv2.Authorization
 	return grants, resp, nil
 }
 
-func readSubjectRoles(subjectID string, authAPI *platformclientv2.AuthorizationApi) (*schema.Set, *platformclientv2.APIResponse, diag.Diagnostics) {
+func readSubjectRoles(d *schema.ResourceData, subjectID string, authAPI *platformclientv2.AuthorizationApi) ([]interface{}, *platformclientv2.APIResponse, diag.Diagnostics) {
 	grants, resp, err := getAssignedGrants(subjectID, authAPI)
 	if err != nil {
 		return nil, resp, err
 	}
 
-	roleDivsMap := make(map[string]*schema.Set)
+	roleDivsMap := make(map[string][]interface{})
 	for _, grant := range grants {
 		if currentDivs, ok := roleDivsMap[*grant.Role.Id]; ok {
-			currentDivs.Add(*grant.Division.Id)
+			currentDivs = append(currentDivs, *grant.Division.Id)
 		} else {
-			roleDivsMap[*grant.Role.Id] = schema.NewSet(schema.HashString, []interface{}{*grant.Division.Id})
+			roleDivsMap[*grant.Role.Id] = []interface{}{*grant.Division.Id}
 		}
 	}
 
-	roleSet := schema.NewSet(schema.HashResource(roleAssignmentResource), []interface{}{})
+	var roleList []interface{}
 	for roleID, divs := range roleDivsMap {
 		role := make(map[string]interface{})
 		role["role_id"] = roleID
 		role["division_ids"] = divs
-		roleSet.Add(role)
+		roleList = append(roleList, role)
 	}
-	return roleSet, resp, nil
+
+	// If the role IDs are the same in the schema state and in the response from the GET,
+	// re-organize the items to match the ordering in the schema
+	rolesFromSchema, ok := d.Get("roles").([]interface{})
+	if !ok {
+		return roleList, resp, nil
+	}
+
+	roleIdsFromSchema := getRoleIdsFromRolesList(rolesFromSchema)
+	var roleIdsFromApi []string
+	for roleId, _ := range roleDivsMap {
+		roleIdsFromApi = append(roleIdsFromApi, roleId)
+	}
+
+	if lists.AreEquivalent(roleIdsFromSchema, roleIdsFromApi) {
+		// re-organise roleList so that order of items is the same as in the schema
+		roleListReordered := make([]interface{}, 0)
+		for _, roleId := range roleIdsFromSchema {
+			currentRole := make(map[string]interface{}, 0)
+			currentRole["role_id"] = roleId
+			currentRole["division_ids"] = roleDivsMap[roleId]
+			roleListReordered = append(roleListReordered, currentRole)
+		}
+		roleList = roleListReordered
+	}
+
+	return roleList, resp, nil
 }
 
-func updateSubjectRoles(ctx context.Context, d *schema.ResourceData, authAPI *platformclientv2.AuthorizationApi, subjectType string) diag.Diagnostics {
-	if d.HasChange("roles") {
-		rolesConfig := d.Get("roles")
```

Comment:
The only real changes to this function was moving everything out of the if statements to reduce nesting

Instead of it being:

```golang
if d.HasChange("roles") {
  // all the code
}
return nil
```

it is:

```golang
if !d.HasChange("roles") {
  return nil
}
// all the code ...
return nil
```